server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -4173,6 +4173,6 @@ public String toString() {
     }
 
     public void waitForSegmentGeneration(long segmentGeneration, ActionListener<Long> listener) {
-        getEngine().addSegmentGenerationListener(segmentGeneration, listener);
+        getEngine().addSegmentGenerationListener(getOperationPrimaryTerm(), segmentGeneration, listener);
```

Comment:
I originally though about adding the primary term to 
```
PostWriteRefresh
TransportShardMultiGetFomTranslogAction#Response
TransportGetFromTranslogAction#Response
```
but I found that they all also would derive the primary term from the primary `IndexShard` that it seemed simpler to just call `operationPrimaryTerm`?

Reply:
This will be the operation primary term on the search shard. That could be different from the primary term used to create the generation. Unless I am missing something, we could be using the wrong primary term?

Reply:
I've moved acquiring of the operation primary term on the index shard in `PostWriteRefresh`, `TransportShardMultiGetFomTranslogAction` and `TransportGetFromTranslogAction`